File: Assets/StreamingAssets/Lua/Logic/LgDestroyer.lua
--
-- The destroyer class.
--
-- @filename  LgDestroyer.lua
-- @copyright Copyright (c) 2015 Yaukey/yaukeywang/WangYaoqi (yaukeywang@gmail.com) all rights reserved.
-- @license   The MIT License (MIT)
-- @author    Yaukey
-- @date      2015-09-05
--

local YwDeclare = YwDeclare
local YwClass = YwClass

local DLog = YwDebug.Log
local DLogWarn = YwDebug.LogWarning
local DLogError = YwDebug.LogError

-- Register new class LgDestroyer.
local strClassName = "LgDestroyer"
local LgDestroyer = YwDeclare(strClassName, YwClass(strClassName))

-- Member variables.

-- The c# class object.
LgDestroyer.this = false

-- The transform.
LgDestroyer.transform = false

-- The c# gameObject.
LgDestroyer.gameObject = false

-- Whether or not this gameobject should destroyed after a delay, on Awake.
LgDestroyer.m_bDestroyOnAwake = false

-- The delay for destroying it on Awake.
LgDestroyer.m_fAwakeDestroyDelay = 0.0

-- Find a child game object and delete it.
LgDestroyer.m_bFindChild = false

-- Name the child object in Inspector.
LgDestroyer.m_strNamedChild = ""

-- Awake method.
function LgDestroyer:Awake()
    --print("LgDestroyer:Awake")

    -- Check variable.
    if (not self.this) or (not self.transform) or (not self.gameObject) then
        DLogError("Init error in LgDestroyer!")
        return
    end

    -- If the gameobject should be destroyed on awake,
    if self.m_bDestroyOnAwake then
        if self.m_bFindChild then
            GameObject.Destroy(self.transform:Find(self.m_strNamedChild).gameObject)
        else
            -- ... destroy the gameobject after the delay.
            GameObject.Destroy(self.gameObject, self.m_fAwakeDestroyDelay)
        end
    end
end

-- Destroy child game object method.
function LgDestroyer:DestroyChildGameObject()
    --print("LgDestroyer:DestroyChildGameObject")

    -- Destroy this child gameobject, this can be called from an Animation Event.
    local cObj = self.transform:Find(self.m_strNamedChild).gameObject
    if not Slua.IsNull(cObj) then
        GameObject.Destroy(cObj)
    end
end

-- Disable child game object method.
function LgDestroyer:DisableChildGameObject()
    --print("LgDestroyer:DisableChildGameObject")

    -- Destroy this child gameobject, this can be called from an Animation Event.
    local cObj = self.transform:Find(self.m_strNamedChild).gameObject
    if cObj.activeSelf then
        cObj:SetActive(false)
    end
end

function LgDestroyer:DestroyGameObject()
    --print("LgDestroyer:DestroyGameObject")

    -- Destroy this gameobject, this can be called from an Animation Event.
    GameObject.Destroy(self.gameObject)
end

-- Return this class.
return LgDestroyer

File: Assets/Game/Scripts/Logic/LgDestroyer.cs
/**
 * Destroyer class.
 *
 * @filename  LgDestroyer.cs
 * @copyright Copyright (c) 2015 Yaukey/yaukeywang/WangYaoqi (yaukeywang@gmail.com) all rights reserved.
 * @license   The MIT License (MIT)
 * @author    Yaukey
 * @date      2015-09-05
 */

using UnityEngine;
using System.Collections;

// The destroyer class.
public class LgDestroyer : MonoBehaviour
{
	// Whether or not this gameobject should destroyed after a delay, on Awake.
	public bool m_destroyOnAwake = false;

	// The delay for destroying it on Awake.
	public float m_awakeDestroyDelay = 0.0f;

	// Find a child game object and delete it.
	public bool m_findChild = false;

	// Name the child object in Inspector.
	public string m_namedChild = string.Empty;
	
	// The awake method.
	void Awake()
	{
        // If the gameobject should be destroyed on awake,
        if (m_destroyOnAwake)
        {
            if (m_findChild)
            {
                Destroy(transform.Find(m_namedChild).gameObject);
            }
            else
            {
                // ... destroy the gameobject after the delay.
                Destroy(gameObject, m_awakeDestroyDelay);
            }
        }
    }

	// Destroy this child gameobject, this can be called from an Animation Event.
	public void DestroyChildGameObject()
	{
        // Destroy this child gameobject, this can be called from an Animation Event.
        if (null!= transform.Find(m_namedChild).gameObject)
        {
            Destroy(transform.Find(m_namedChild).gameObject);
        }
    }

	// Destroy this child gameobject, this can be called from an Animation Event.
	public void DisableChildGameObject()
	{
        // Destroy this child gameobject, this can be called from an Animation Event.
        if (transform.Find(m_namedChild).gameObject.activeSelf)
        {
            transform.Find(m_namedChild).gameObject.SetActive(false);
        }
    }

	// Destroy this gameobject, this can be called from an Animation Event.
	public void DestroyGameObject()
	{
        // Destroy this gameobject, this can be called from an Animation Event.
        Destroy(gameObject);
    }
}
